crates/ruff_linter/src/rules/flake8_pyi/rules/duplicate_union_member.rs

```diff
@@ -27,63 +31,176 @@ use crate::checkers::ast::Checker;
     }
 }
 
 /// PYI016
 pub(crate) fn duplicate_union_member<'a>(checker: &mut Checker, expr: &'a Expr) {
     let mut seen_nodes: HashSet<ComparableExpr<'_>, _> = FxHashSet::default();
+    let mut unique_nodes: Vec<&Expr> = Vec::new();
     let mut diagnostics: Vec<Diagnostic> = Vec::new();
 
+    let mut union_type = UnionKind::TypingUnion;
     // Adds a member to `literal_exprs` if it is a `Literal` annotation
     let mut check_for_duplicate_members = |expr: &'a Expr, parent: &'a Expr| {
+        if matches!(parent, Expr::BinOp(_)) {
+            union_type = UnionKind::PEP604;
+        }
+
         // If we've already seen this union member, raise a violation.
-        if !seen_nodes.insert(expr.into()) {
-            let mut diagnostic = Diagnostic::new(
+        if seen_nodes.insert(expr.into()) {
+            unique_nodes.push(expr);
+        } else {
+            diagnostics.push(Diagnostic::new(
                 DuplicateUnionMember {
                     duplicate_name: checker.generator().expr(expr),
                 },
                 expr.range(),
-            );
-            // Delete the "|" character as well as the duplicate value by reconstructing the
-            // parent without the duplicate.
-
-            // If the parent node is not a `BinOp` we will not perform a fix
-            if let Expr::BinOp(ast::ExprBinOp { left, right, .. }) = parent {
-                // Replace the parent with its non-duplicate child.
-                let child = if expr == left.as_ref() { right } else { left };
-                diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
-                    checker.locator().slice(child.as_ref()).to_string(),
-                    parent.range(),
-                )));
-            }
-            diagnostics.push(diagnostic);
+            ));
         }
     };
 
     // Traverse the union, collect all diagnostic members
     traverse_union(&mut check_for_duplicate_members, checker.semantic(), expr);
 
+    if diagnostics.is_empty() {
+        return;
+    }
+
+    // Mark [`Fix`] as unsafe when comments are in range.
+    let applicability = if checker.comment_ranges().intersects(expr.range()) {
+        Applicability::Unsafe
+    } else {
+        Applicability::Safe
+    };
+
+    // Generate the flattened fix once.
+    let fix = if let &[edit_expr] = unique_nodes.as_slice() {
+        // Generate a [`Fix`] for a single type expression, e.g. `int`.
+        Fix::applicable_edit(
+            Edit::range_replacement(checker.generator().expr(edit_expr), expr.range()),
+            applicability,
+        )
+    } else {
+        match union_type {
+            // See redundant numeric union
+            UnionKind::PEP604 => generate_pep604_fix(checker, unique_nodes, expr, applicability),
+            UnionKind::TypingUnion => {
+                generate_union_fix(checker, unique_nodes, expr, applicability)
+                    .ok()
```

Comment:
I haven't looked into in which cases the import request errors (and probably should handle it differently)

Reply:
Yes, we can't `unwrap` here and it will mean that we can't mark this rule as always fixable.